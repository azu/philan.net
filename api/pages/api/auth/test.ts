import { NextApiResponse } from "next";
import nextConnect from "next-connect";
import { NextApiRequestWithSession, withSession } from "../../../api-utils/with-session";
import { randomBytes } from "crypto";

const createRandom = () => {
    return randomBytes(40).reduce((p, i) => p + (i % 32).toString(32), "");
};

const handler = nextConnect<NextApiRequestWithSession, NextApiResponse>()
    .use(withSession())
    .get(async (req, res) => {
        const uuid = createRandom();
        req.session.authState = uuid;
        res.json({
            uuid
        });
    });

export default handler;
