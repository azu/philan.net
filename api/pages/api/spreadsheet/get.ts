import { NextApiResponse } from "next";
import { google } from "googleapis";
import { UserCredentials } from "../../../domain/User";
import { createOAuthClient } from "../../../api-utils/create-OAuth";
import nextConnect from "next-connect";
import { withSession } from "../../../api-utils/with-session";
import { NextApiRequestWithUserSession, requireLogin } from "../../../api-utils/requireLogin";

const sheets = google.sheets("v4");
export const getSpreadSheet = async ({
    spreadsheetId,
    credentials
}: {
    spreadsheetId: string;
    credentials: UserCredentials;
}) => {
    const client = createOAuthClient(credentials);
    const { token } = await client.getAccessToken();
    if (!token) {
        throw new Error("No Access Token");
    }
    const spreadsheet = await sheets.spreadsheets.get({
        oauth_token: token,
        spreadsheetId,
        includeGridData: true
    });
    const START_OF_USER_DATA = 3;
    if (!spreadsheet?.data?.sheets) {
        throw new Error("500: No spreadsheet data");
    }
    return spreadsheet?.data?.sheets?.map((sheet) => {
        const items = sheet?.data?.[0];
        // ["Budget", "Used", "Balance"]
        const statsRow = items?.rowData?.[1];
        const rowDate = items?.rowData?.slice(START_OF_USER_DATA) ?? [];
        return {
            year: sheet?.properties?.title!,
            README: statsRow?.values?.[3]?.formattedValue!,
            stats: {
                budget: {
                    raw: statsRow?.values?.[0]?.effectiveValue?.numberValue!,
                    value: statsRow?.values?.[0]?.formattedValue!
                },
                used: {
                    raw: statsRow?.values?.[1]?.effectiveValue?.numberValue!,
                    value: statsRow?.values?.[1]?.formattedValue!
                },
                balance: {
                    raw: statsRow?.values?.[2]?.effectiveValue?.numberValue!,
                    value: statsRow?.values?.[2]?.formattedValue!
                }
            },
            items: rowDate
                .filter((row) => {
                    const values = row.values;
                    return (
                        values?.[0].userEnteredValue?.stringValue &&
                        values?.[1].userEnteredValue?.stringValue &&
                        values?.[2]?.formattedValue!
                    );
                })
                ?.map((row) => {
                    const values = row.values;
                    return {
                        date: values?.[0].userEnteredValue?.stringValue!,
                        to: values?.[1].userEnteredValue?.stringValue!,
                        amount: {
                            raw: values?.[2].userEnteredValue?.numberValue!,
                            value: values?.[2]?.formattedValue!
                        },
                        url: values?.[3].userEnteredValue?.stringValue!,
                        memo: values?.[4].userEnteredValue?.stringValue ?? ""
                    };
                })
        };
    });
};
const handler = nextConnect<NextApiRequestWithUserSession, NextApiResponse>()
    .use(withSession())
    .use(requireLogin())
    .post(async (req, res) => {
        const user = req.user;
        const response = await getSpreadSheet({
            credentials: user.credentials,
            spreadsheetId: user.spreadsheetId
        });
        res.json(response);
    });

export default handler;
