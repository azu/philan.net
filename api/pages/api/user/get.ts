import nextConnect from "next-connect";
import { NextApiRequestWithSession, withSession } from "../../../api-utils/with-session";
import { NextApiResponse } from "next";
import { createUserKvs } from "../../../api-utils/userKvs";
import { logger } from "../../../api-utils/logger";

const handler = nextConnect<NextApiRequestWithSession, NextApiResponse>()
    .use(withSession())
    .get(async (req, res) => {
        const userKVS = createUserKvs();
        logger.debug("Google UserId: %s", req.session.get("googleUserId"));
        const user = await userKVS.findByGoogleId(req.session.get("googleUserId"));
        if (!user) {
            return res.send({
                login: false
            });
        }
        res.json({
            login: true
        });
    });
export default handler;
