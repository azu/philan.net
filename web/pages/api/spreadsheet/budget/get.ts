import { NextApiResponse } from "next";
import { google, sheets_v4 } from "googleapis";
import { UserCredentials } from "../../../../domain/User";
import { createOAuthClient } from "../../../../api-utils/create-OAuth";
import nextConnect from "next-connect";
import { withSession } from "../../../../api-utils/with-session";
import { NextApiRequestWithUserSession, requireLogin } from "../../../../api-utils/requireLogin";
import { BudgetItem, GetBudgetResponse } from "./api-types";
import { SheetTitles } from "../SpreadSheetSchema";
import { isBudgetItem } from "./api-types.validator";
type Schema$Sheet = sheets_v4.Schema$Sheet;

const sheets = google.sheets("v4");
export const parseBudgetsFromBudgetsSheet = (budgetsSheet: Schema$Sheet): BudgetItem[] => {
    const START_OF_USER_DATA = 1;
    const recordAllCells = budgetsSheet?.data?.[0].rowData;
    const recordDataCells = recordAllCells?.slice(START_OF_USER_DATA) ?? [];
    return recordDataCells
        .filter((cell) => {
            return cell.values !== undefined;
        })
        .map((cell) => {
            // Year	Budget	Used	Balance
            const [Year, Budget, Used, Balance] = cell.values!;
            return {
                year: Number(Year.userEnteredValue?.stringValue),
                budget: {
                    raw: Budget.effectiveValue?.numberValue,
                    value: Budget.formattedValue
                },
                used: {
                    raw: Used.effectiveValue?.numberValue,
                    value: Used.formattedValue
                },
                balance: {
                    raw: Balance.effectiveValue?.numberValue,
                    value: Balance.formattedValue
                }
            };
        })
        .filter((item) => {
            return isBudgetItem(item);
        }) as BudgetItem[];
};
export const getBudgets = async ({
    spreadsheetId,
    credentials
}: {
    spreadsheetId: string;
    credentials: UserCredentials;
    defaultCurrency: string;
}): Promise<GetBudgetResponse> => {
    const client = createOAuthClient(credentials);
    const { token } = await client.getAccessToken();
    if (!token) {
        throw new Error("No Access Token");
    }
    const spreadsheet = await sheets.spreadsheets.get({
        oauth_token: token,
        spreadsheetId,
        includeGridData: true
    });
    if (!spreadsheet?.data?.sheets) {
        throw new Error("500: No spreadsheet data");
    }
    // const locale = spreadsheet.data.properties?.locale;
    const budgetsSheet = spreadsheet?.data?.sheets?.find((sheet) => {
        return sheet?.properties?.title === SheetTitles.Budgets;
    });
    if (!budgetsSheet) {
        throw new Error(
            "Records Spreadsheet is not found.\n" + "\n" + "Can you rename record spreadsheet to 'Records'?"
        );
    }
    return parseBudgetsFromBudgetsSheet(budgetsSheet);
};

const handler = nextConnect<NextApiRequestWithUserSession, NextApiResponse>()
    .use(withSession())
    .use(requireLogin())
    .get(async (req, res) => {
        const user = req.user;
        const response = await getBudgets({
            credentials: user.credentials,
            spreadsheetId: user.spreadsheetId,
            defaultCurrency: user.defaultCurrency
        });
        res.json(response);
    });

export default handler;
