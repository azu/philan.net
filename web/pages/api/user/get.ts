import nextConnect from "next-connect";
import { NextApiRequestWithSession, withSession } from "../../../api-utils/with-session";
import { NextApiResponse } from "next";
import { createUserKvs } from "../../../api-utils/userKvs";
import { GetUserResponseBody } from "./api-types";

const handler = nextConnect<NextApiRequestWithSession, NextApiResponse<GetUserResponseBody>>()
    .use(withSession())
    .get(async (req, res) => {
        const userKVS = await createUserKvs();
        const user = await userKVS.findByGoogleId(req.session.get("googleUserId"));
        if (!user) {
            return res.send({
                isLogin: false
            });
        }
        res.json({
            isLogin: true,
            id: user.id,
            name: user.name,
            defaultCurrency: user.defaultCurrency,
            avatarUrl: user.avatarUrl,
            spreadsheetUrl: `https://docs.google.com/spreadsheets/d/${user.spreadsheetId}`
        });
    });
export default handler;
