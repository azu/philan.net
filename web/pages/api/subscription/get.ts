import { google } from "googleapis";
import { SheetTitles } from "../spreadsheet/SpreadSheetSchema";
import { NextApiRequestWithUserSession, requireLogin } from "../../../api-utils/requireLogin";
import { NextApiResponse } from "next";
import { withSession } from "../../../api-utils/with-session";
import nextConnect, { ErrorHandler } from "next-connect";
import { getToken, GetTokenMeta } from "../../../api-utils/oauth/getToken";
import dayjs from "dayjs";
import { parseAmount } from "../../../api-utils/parseAmount";
import { logger } from "../../../api-utils/logger";
import type { GetResponseBody, SubscriptionItem } from "./api-types";

export const getSubscriptionItems = async ({
    spreadsheetId,
    defaultCurrency,
    meta
}: {
    spreadsheetId: string;
    defaultCurrency: string;
    meta: GetTokenMeta;
}): Promise<SubscriptionItem[]> => {
    const sheets = google.sheets("v4");
    const token = await getToken(meta);
    if (!token) {
        throw new Error("No Access Token");
    }
    const spreadsheet = await sheets.spreadsheets.get({
        oauth_token: token,
        spreadsheetId,
        includeGridData: true
    });
    if (!spreadsheet?.data?.sheets) {
        throw new Error("500: No spreadsheet data");
    }
    const subscriptionSheet = spreadsheet?.data?.sheets?.find((sheet) => {
        return sheet?.properties?.title === SheetTitles.Subscriptions;
    });
    if (!subscriptionSheet) {
        throw new Error(
            "Records Spreadsheet is not found.\n" + "\n" + "Can you rename record spreadsheet to 'Records'?"
        );
    }
    const START_OF_USER_DATA = 1;
    const recordAllCells = subscriptionSheet?.data?.[0].rowData;
    const recordDataCells = recordAllCells?.slice(START_OF_USER_DATA) ?? [];
    const recordItems = recordDataCells
        .filter((row) => {
            const values = row.values;
            const startDate = values?.[0]?.formattedValue;
            return startDate;
        })
        ?.map((row) => {
            const values = row.values;
            const startDate = values?.[0]?.formattedValue!;
            const eneDate = values?.[1]?.formattedValue ?? "";
            const cron = values?.[2]?.userEnteredValue?.stringValue ?? "";
            const to = values?.[3]?.userEnteredValue?.stringValue ?? "";
            const amountUserEnteredValue = values?.[4]?.userEnteredValue?.formulaValue ?? values?.[4]?.formattedValue;
            const amountNumber = values?.[4]?.effectiveValue?.numberValue!;
            const amountFormattedValue = values?.[4]?.formattedValue!;
            const url = values?.[5]?.userEnteredValue?.stringValue ?? "";
            const memo = values?.[6]?.userEnteredValue?.stringValue ?? "";
            const parsedAmount = parseAmount(amountUserEnteredValue ?? amountNumber, defaultCurrency);
            return {
                startDate,
                eneDate,
                cron,
                to,
                amount: {
                    number: amountNumber,
                    value: amountFormattedValue,
                    raw: parsedAmount.amount,
                    inputCurrency: parsedAmount.from,
                    outputCurrency: parsedAmount.to
                },
                url: url,
                memo: memo
            };
        })
        .sort((a, b) => {
            return dayjs(a.startDate).isBefore(b.startDate) ? 1 : -1;
        });
    return recordItems;
};

const onError: ErrorHandler<any, NextApiResponse> = (error, _req, res) => {
    logger.error(error);
    if (process.env.NODE_ENV === "production") {
        res.status(500).end("Server Error");
    } else {
        res.status(500).end(error.stack);
    }
};
const handler = nextConnect<NextApiRequestWithUserSession, NextApiResponse<GetResponseBody>>({ onError })
    .use(withSession())
    .use(requireLogin())
    .get(async (req, res) => {
        const user = req.user;
        if (!user) {
            throw new Error("No user");
        }
        if (!user.appsScriptId) {
            throw new Error("Already created subscription apps and sheet!");
        }
        const items = await getSubscriptionItems({
            spreadsheetId: user.spreadsheetId,
            defaultCurrency: user.defaultCurrency,
            meta: {
                credentials: user.credentials
            }
        });
        res.json({
            items: items
        });
    });
export default handler;
