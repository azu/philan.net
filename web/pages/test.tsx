import { useEffect, useMemo } from "react";
import { useLocalStorage } from "react-use";
import * as SpreadSheetAPI from "./api/spreadsheet/api-types";
import dynamic from "next/dynamic";

function useSpreadSheet(token?: string) {
    const [spreadsheetId, setspreadsheetId] = useLocalStorage<string>("spreadsheetId", "");
    const handlers = useMemo(
        () => ({
            create: () => {
                const param = new URLSearchParams([["token", token!]]);
                return fetch("/api/spreadsheet/create?" + param, {
                    method: "post",
                    headers: {
                        "Content-Type": "application/json"
                    }
                })
                    .then((res) => {
                        return res.json();
                    })
                    .then((res) => {
                        setspreadsheetId(res.spreadsheetId);
                    });
            },
            add: () => {
                const param = new URLSearchParams([
                    ["token", token!],
                    ["spreadsheetId", spreadsheetId!]
                ]);
                return fetch("/api/spreadsheet/add?" + param, {
                    method: "post",
                    headers: {
                        "Content-Type": "application/json"
                    },
                    body: JSON.stringify({
                        to: "philan.net",
                        amount: 100,
                        url: "https://philan.net",
                        memo: "THIS IS MEMO"
                    } as SpreadSheetAPI.AddRequestBody)
                });
            },
            get: () => {
                const param = new URLSearchParams([
                    ["token", token!],
                    ["spreadsheetId", spreadsheetId!]
                ]);
                return fetch("/api/spreadsheet/get?" + param)
                    .then((res) => {
                        return res.json();
                    })
                    .then((res) => {
                        console.log(res);
                    });
            }
        }),
        [token, spreadsheetId]
    );

    return [spreadsheetId, handlers] as const;
}

const DebugPage = () => {
    const [token, setToken, removeToken] = useLocalStorage<string>("token", "");
    const [spreadsheetId, handlers] = useSpreadSheet(token);
    const authorize = () => {
        removeToken();
        location.href = "/api/auth?state=test";
    };
    useEffect(() => {
        const code = new URL(window.location.href).searchParams.get("code");
        if (!code) {
            return;
        }
        if (token) {
            return;
        }
        (async function () {
            const res = await fetch("/api/auth/getToken?code=" + code).then((res) => res.json());
            const access_token: string = res.tokens.access_token;
            setToken(access_token);
            localStorage.setItem("access_token", access_token);
        })();
    }, []);
    if (token) {
        return (
            <div>
                <h2>SpredSheet</h2>
                <p>
                    : <a href={`https://docs.google.com/spreadsheets/d/${spreadsheetId}/`}>{spreadsheetId}</a>
                </p>
                <ul>
                    <li>
                        <button onClick={handlers.create}>Create</button>
                    </li>
                    <li>
                        <button onClick={handlers.add}>Add</button>
                    </li>

                    <li>
                        <button onClick={handlers.get}>Get</button>
                    </li>
                </ul>
                <h2>Page</h2>
                <ul>
                    <li>
                        <a href={`/user/${spreadsheetId}/${token}`}>User</a>
                    </li>
                </ul>
                <footer>
                    <label>
                        Re-Authorize: <button onClick={authorize}>Authorize</button>
                    </label>
                </footer>
            </div>
        );
    } else {
        return (
            <div>
                <p>
                    <a href={"/api/auth?state=test"}>Authorize</a>
                </p>
            </div>
        );
    }
};
export default dynamic(
    async () => {
        if (process.env.NODE_ENV !== "development") {
            return () => null;
        }
        return DebugPage;
    },
    { loading: () => <p>...</p> }
);
