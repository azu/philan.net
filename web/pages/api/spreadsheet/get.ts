import { NextApiResponse } from "next";
import { google } from "googleapis";
import { UserCredentials } from "../../../domain/User";
import { createOAuthClient } from "../../../api-utils/create-OAuth";
import nextConnect from "next-connect";
import { withSession } from "../../../api-utils/with-session";
import { NextApiRequestWithUserSession, requireLogin } from "../../../api-utils/requireLogin";
import dayjs from "dayjs";
import { createItemId } from "../../../api-utils/create-item-id";
import { GetResponseBody } from "./api-types";
import groupBy from "lodash/groupBy";
import { SheetTitles } from "./SpreadSheetSchema";
import { parseBudgetsFromBudgetsSheet } from "./budget/get";

const sheets = google.sheets("v4");

const parseAmount = (amount: string | number, defaultCurrency: string) => {
    if (amount === undefined) {
        throw new Error("amount is undefined. Please check your spreadsheet");
    }
    if (typeof amount === "number") {
        return {
            amount,
            from: defaultCurrency,
            to: defaultCurrency
        };
    }
    // = ${v} * index(GOOGLEFINANCE("CURRENCY:${item.currency.from}${item.currency.to}", "price", "${date}"), 2, 2);
    if (amount.startsWith("=")) {
        const match = amount.match(
            /=\s*(?<AMOUNT>\d+)\s*\*\s*index\(GOOGLEFINANCE\("CURRENCY:(?<FROM>\w{3})(?<TO>\w{3})/
        );
        if (!match) {
            return {
                amount: 0,
                from: defaultCurrency,
                to: defaultCurrency
            };
        }
        return {
            amount: Number(match.groups?.AMOUNT!),
            from: match.groups?.FROM!,
            to: match.groups?.TO!
        };
    } else {
        return {
            amount: Number(amount),
            from: defaultCurrency,
            to: defaultCurrency
        };
    }
};
export const getSpreadSheet = async ({
    spreadsheetId,
    credentials,
    defaultCurrency
}: {
    spreadsheetId: string;
    credentials: UserCredentials;
    defaultCurrency: string;
}): Promise<GetResponseBody> => {
    const client = createOAuthClient(credentials);
    const { token } = await client.getAccessToken();
    if (!token) {
        throw new Error("No Access Token");
    }
    const spreadsheet = await sheets.spreadsheets.get({
        oauth_token: token,
        spreadsheetId,
        includeGridData: true
    });
    const START_OF_USER_DATA = 3;
    if (!spreadsheet?.data?.sheets) {
        throw new Error("500: No spreadsheet data");
    }
    // const locale = spreadsheet.data.properties?.locale;
    const recordSheet = spreadsheet?.data?.sheets?.find((sheet) => {
        // TODO: OLD_Records is for backward-compatible
        return sheet?.properties?.title === SheetTitles.Records || sheet?.properties?.title === SheetTitles.OLD_Records;
    });
    const budgetsSheet = spreadsheet?.data?.sheets?.find((sheet) => {
        return sheet?.properties?.title === SheetTitles.Budgets;
    });
    if (!recordSheet) {
        throw new Error(
            "Records Spreadsheet is not found.\n" + "\n" + "Your spreadsheet does not have 'Records' sheet"
        );
    }
    if (!budgetsSheet) {
        throw new Error(
            "Budgets Spreadsheet is not found.\n" + "\n" + "Your spreadsheet does not have 'Budgets' sheet?"
        );
    }
    const recordAllCells = recordSheet?.data?.[0].rowData;
    const recordDataCells = recordAllCells?.slice(START_OF_USER_DATA) ?? [];
    const recordItems = recordDataCells
        .filter((row) => {
            const values = row.values;
            return (
                values?.[0].userEnteredValue?.stringValue &&
                values?.[1].userEnteredValue?.stringValue &&
                values?.[2]?.formattedValue!
            );
        })
        ?.map((row) => {
            const values = row.values;
            const dateString = values?.[0]?.userEnteredValue?.stringValue!;
            const amountUserEnteredValue = values?.[2]?.userEnteredValue?.formulaValue;
            const amountNumber = values?.[2]?.effectiveValue?.numberValue!;
            const amountFormattedValue = values?.[2]?.formattedValue!;
            const url = values?.[3]?.userEnteredValue?.stringValue!;
            const id = createItemId({
                dateString,
                amountNumber,
                url
            });
            const meta = JSON.parse(values?.[5]?.userEnteredValue?.stringValue ?? "{}");
            const parsedAmount = parseAmount(amountUserEnteredValue ?? amountNumber, defaultCurrency);
            return {
                id,
                date: dateString,
                to: values?.[1].userEnteredValue?.stringValue!,
                amount: {
                    number: amountNumber,
                    value: amountFormattedValue,
                    raw: parsedAmount.amount,
                    inputCurrency: parsedAmount.from,
                    outputCurrency: parsedAmount.to
                },
                url: url ?? "",
                memo: values?.[4].userEnteredValue?.stringValue ?? "",
                meta: meta
            };
        })
        .sort((a, b) => {
            return dayjs(a.date).isBefore(b.date) ? 1 : -1;
        });
    const statsRow = recordAllCells?.[1];
    if (!statsRow) {
        throw new Error("statsRow is not defined");
    }
    const README = statsRow?.values?.[3]?.formattedValue! ?? "";
    const budgets = parseBudgetsFromBudgetsSheet(budgetsSheet);
    const matchBudget = (year: number) => {
        const matchedBudget = budgets.find((budget) => budget.year === year);
        if (matchedBudget) {
            return matchedBudget;
        }
        return {
            year: year,
            budget: {
                raw: 0,
                value: "None"
            },
            used: {
                raw: 0,
                value: "None"
            },
            balance: {
                raw: 0,
                value: "None"
            }
        };
    };
    if (recordItems.length === 0) {
        const currentYear = dayjs().format("YYYY");
        const budget = matchBudget(Number(currentYear));
        return [
            {
                year: currentYear,
                stats: budget,
                README,
                items: []
            }
        ];
    }
    const itemsByYear = groupBy(recordItems, (item) => dayjs(item.date).format("YYYY"));
    // 2021, 2020 ....
    return Object.keys(itemsByYear)
        .sort((a, b) => Number(b) - Number(a))
        .map((year) => {
            const budget = matchBudget(Number(year));
            return {
                year,
                stats: budget,
                README,
                items: itemsByYear[year]
            };
        });
};

const handler = nextConnect<NextApiRequestWithUserSession, NextApiResponse>()
    .use(withSession())
    .use(requireLogin())
    .get(async (req, res) => {
        const user = req.user;
        const response = await getSpreadSheet({
            credentials: user.credentials,
            spreadsheetId: user.spreadsheetId,
            defaultCurrency: user.defaultCurrency
        });
        res.json(response);
    });

export default handler;
