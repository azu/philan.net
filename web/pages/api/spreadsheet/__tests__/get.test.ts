import { getTestCredential, hasTestCredential, TEST_SPREADSHEET_ID } from "../../../../test/GoogleCredential";
import { getSpreadSheet } from "../get";

const describe = hasTestCredential() ? global.describe : global.describe.skip;
describe("get", () => {
    it("snapshot testing", async () => {
        const result = await getSpreadSheet({
            spreadsheetId: TEST_SPREADSHEET_ID,
            credentials: getTestCredential(),
            defaultCurrency: "JPY"
        });
        expect(result).toMatchSnapshot();
    });
});
