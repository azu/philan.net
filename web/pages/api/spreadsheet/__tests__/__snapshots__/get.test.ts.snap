// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get snapshot testing 1`] = `
Array [
  Object {
    "README": "SPREADSHEET FOR TESTING",
    "items": Array [
      Object {
        "amount": Object {
          "inputCurrency": "USD",
          "number": 108.553,
          "outputCurrency": "JPY",
          "raw": 1,
          "value": "¥108.55",
        },
        "date": "2021-03-11T14:19:57.900Z",
        "id": "2021-03-11-Z2lx9rw",
        "memo": "Memo 3",
        "meta": Object {
          "id": "2021-03-11-ZOUAwW",
          "type": "checked",
        },
        "to": "USD TO",
        "url": "https://example.com/C",
      },
      Object {
        "amount": Object {
          "inputCurrency": "JPY",
          "number": 100,
          "outputCurrency": "JPY",
          "raw": 100,
          "value": "100",
        },
        "date": "2021-02-14T08:37:13.194Z",
        "id": "2021-02-14-ZS9okG",
        "memo": "Memo 2",
        "meta": Object {
          "type": "checked",
        },
        "to": "to philan",
        "url": "https://philan.net",
      },
      Object {
        "amount": Object {
          "inputCurrency": "JPY",
          "number": 100,
          "outputCurrency": "JPY",
          "raw": 100,
          "value": "100",
        },
        "date": "2021-02-14T08:07:07.947Z",
        "id": "2021-02-14-Z1gUQfD",
        "memo": "Memo 1",
        "meta": Object {
          "type": "checked",
        },
        "to": "寄付先",
        "url": "https://example.com/A",
      },
    ],
    "stats": Object {
      "balance": Object {
        "raw": 99691.447,
        "value": "¥99,691.45",
      },
      "budget": Object {
        "raw": 100000,
        "value": "¥100,000.00",
      },
      "used": Object {
        "raw": 308.553,
        "value": "¥308.55",
      },
    },
    "year": "2021",
  },
  Object {
    "README": "SPREADSHEET FOR TESTING",
    "items": Array [
      Object {
        "amount": Object {
          "inputCurrency": "JPY",
          "number": 2020,
          "outputCurrency": "JPY",
          "raw": 2020,
          "value": "2020",
        },
        "date": "2020-12-01T08:07:07.947Z",
        "id": "2020-12-01-Z14uIfN",
        "memo": "Memo 2 @ 2020",
        "meta": Object {
          "type": "checked",
        },
        "to": "2020 donation 2",
        "url": "https://example.com/",
      },
      Object {
        "amount": Object {
          "inputCurrency": "JPY",
          "number": 2020,
          "outputCurrency": "JPY",
          "raw": 2020,
          "value": "2020",
        },
        "date": "2020-01-01T08:07:07.947Z",
        "id": "2020-01-01-Z20nzFG",
        "memo": "Memo 1 @ 2020",
        "meta": Object {
          "type": "checked",
        },
        "to": "2020 donation 1",
        "url": "https://example.com/",
      },
    ],
    "stats": Object {
      "balance": Object {
        "raw": 99691.447,
        "value": "¥99,691.45",
      },
      "budget": Object {
        "raw": 100000,
        "value": "¥100,000.00",
      },
      "used": Object {
        "raw": 308.553,
        "value": "¥308.55",
      },
    },
    "year": "2020",
  },
]
`;
